### PR TITLE
Feature/move first car to next section

### PR DESCRIPTION
**Pull request by ktr2884**
added  queue in streetsection with getter and setter

### CODE REVIEW

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/Car.java [hunk 1]
```diff
@@ -3,4 +3,6 @@
 import at.fhv.itm3.s2.roundabout.api.entity.ICar;
 
 public class Car implements ICar {
+
+
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/Car.java]: delete spaces or maybe set auto format on commit (don't know if this helps)

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 2]
```diff
@@ -8,8 +8,11 @@
 
 import java.util.Map;
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: delete space

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 3]
```diff
@@ -40,6 +43,11 @@ public IStreetConnector getPreviousStreetConnector() {
         return null;
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: why return null in a constructor?

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 4]
```diff
@@ -40,6 +43,11 @@ public IStreetConnector getPreviousStreetConnector() {
...
     }
 
+    @Override
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: why override? it is only a interface method implementation

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 5]
```diff
@@ -56,8 +64,11 @@ public void moveFirstCarToNextSection() {
 
     }
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: where is this bracket from?

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 6]
```diff
@@ -56,8 +64,11 @@ public void moveFirstCarToNextSection() {
...
 
-    @Override
-    public void addCar(ICar car) {
+    public Queue<Car> getQueue() {
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: i dont think it should be possible to manipulate the queue from outside, this is a hidden attribute of the section

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 7]
```diff
@@ -56,8 +64,11 @@ public void moveFirstCarToNextSection() {
...
+        return queue;
+    }
 
+    public void setQueue(Queue<Car> queue) {
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: queue should not be set from outside, it should be managed by the section only

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java (hunk 6 above)
Comment (Oxydation) [reply roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: i think the same, the user should not know what data structure is behind the street section

**Comment by DaGrisa** [comment]
features already implemented by @Oxydation, i think this PR can be closed

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java (hunk 7 above)
Comment (sereGkaluv) [reply roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: +1 consider initializing queue within constructor, also refactor holding property to be **final**.

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 8]
```diff
@@ -66,12 +66,31 @@ public boolean carCouldEnterNextSection() {
...
+     * queue of the next streetSection of the route of the car, if there is one.
+     * If the current streetSection was the last one of the route the car disappears in a sink.
+     */
     public void moveFirstCarToNextSection() {
-        throw new NotImplementedException();
+        Car firstCar = (Car)carQueue.poll();
+        if (firstCar!=null){
+            StreetSection currentSection = firstCar.getCurrentSection();
+            int currentSectionIndex = firstCar.getRoute().indexOf(currentSection);
+            if (currentSectionIndex < (firstCar.getRoute().size()-1)){
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: you could use getDestination() from car, to get the destination section
just define the method in ICar interface

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 9]
```diff
@@ -66,12 +66,31 @@ public boolean carCouldEnterNextSection() {
...
+                StreetSection nextSection = firstCar.getRoute().get(currentSectionIndex+1);
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: you could use getNextStreetSection() from ICar

**Comment by DaGrisa** [comment]
Tests fehlen noch, siehe mein PR

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java (hunk 8 above)
Comment (ktr2884) [reply roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: What is the differnce between getDestination and getNextStreetSection?

Comment (DaGrisa) [reply roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: getDestination always returns the last section in the cars route

File: roundabout-api/src/main/java/at/fhv/itm3/s2/roundabout/api/entity/ICar.java [hunk 10]
```diff
@@ -1,4 +1,30 @@
...
 public interface ICar {
+    /**
+     * returns the index of currentStreetSection in Route
+     */
+    int getCurrentIndexOfRoute();
+
+    /**
+     * returns the route of the car, which is a list<StreetSection>
+     */
+    public List<IStreetSection> getRoute();
```
Comment (DaGrisa) [inline roundabout-api/src/main/java/at/fhv/itm3/s2/roundabout/api/entity/ICar.java]: interface methoden sind immer public

Comment (DaGrisa) [reply roundabout-api/src/main/java/at/fhv/itm3/s2/roundabout/api/entity/ICar.java]: wofür wird getRoute benötigt?

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 11]
```diff
@@ -66,12 +66,41 @@ public boolean carCouldEnterNextSection() {
...
+    /**
+     * moveFirstCarToNextSection removes the first car from the queue and puts it into the
+     * queue of the next streetSection of the route of the car, if there is one.
+     * If the current streetSection was the last one of the route the car disappears in a sink.
+     */
     public void moveFirstCarToNextSection() {
-        throw new NotImplementedException();
+        ICar firstCar = removeFirstCar();
+        if (firstCar!=null){
+            //StreetSection currentSection = firstCar.getCurrentSection();
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: löschen, nicht auskommentieren

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 12]
```diff
@@ -66,12 +66,41 @@ public boolean carCouldEnterNextSection() {
...
+            int currentSectionIndex = firstCar.getCurrentIndexOfRoute();
+            if (currentSectionIndex < (firstCar.getRoute().size()-1)){
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: hier kannst du getDestination verwenden

Comment (DaGrisa) [reply roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: hier kannst du getDestination verwenden

File: roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java [hunk 13]
```diff
@@ -66,12 +66,41 @@ public boolean carCouldEnterNextSection() {
...
+        carQueue.add(car);
+    }
+
+    /**
+     * removes the first car of the queue and returns the first Car
+     * @return
+     */
+    @Override
+    public ICar removeFirstCar() {
+
```
Comment (DaGrisa) [inline roundabout-core/src/main/java/at/fhv/itm3/s2/roundabout/entity/StreetSection.java]: hier kommt die Zeile ein, die vorher oben war ... carQueue.poll()

**Pull request merged**